pulsar-common/src/main/java/com/yahoo/pulsar/common/util/collections/ConcurrentLongPairSet.java

```diff
@@ -0,0 +1,524 @@
+            } finally {
+                if (acquiredLock) {
+                    unlockRead(stamp);
+                }
+            }
+        }
+
+        boolean add(long item1, long item2, long hash) {
+            long stamp = writeLock();
+            int bucket = signSafeMod(hash, capacity);
+
+            // Remember where we find the first available spot
+            int firstDeletedItem = -1;
+
+            try {
+                while (true) {
+                    long storedItem1 = table[bucket];
+                    long storedItem2 = table[bucket + 1];
+
+                    if (item1 == storedItem1 && item2 == storedItem2) {
+                        // Item was already in set
+                        return false;
+                    } else if (storedItem1 == EmptyItem) {
+                        // Found an empty bucket. This means the key is not in the set. If we've already seen a deleted
+                        // key, we should write at that position
+                        if (firstDeletedItem != -1) {
+                            bucket = firstDeletedItem;
+                        } else {
+                            ++usedBuckets;
+                        }
+
+                        table[bucket] = item1;
+                        table[bucket + 1] = item2;
+                        ++size;
+                        return true;
+                    } else if (storedItem1 == DeletedItem) {
+                        // The bucket contained a different deleted key
+                        if (firstDeletedItem == -1) {
+                            firstDeletedItem = bucket;
+                        }
+                    }
+
+                    bucket = (bucket + 2) & (table.length - 1);
+                }
+            } finally {
+                if (usedBuckets > resizeThreshold) {
+                    try {
+                        rehash();
+                    } finally {
+                        unlockWrite(stamp);
+                    }
+                } else {
+                    unlockWrite(stamp);
+                }
+            }
+        }
+
+        private boolean remove(long item1, long item2, int hash) {
+            long stamp = writeLock();
+            int bucket = signSafeMod(hash, capacity);
+
+            try {
+                while (true) {
+                    long storedItem1 = table[bucket];
+                    long storedItem2 = table[bucket + 1];
+                    if (item1 == storedItem1 && item2 == storedItem2) {
+                        --size;
+
+                        cleanBucket(bucket);
+                        return true;
+
+                    } else if (storedItem1 == EmptyItem) {
+                        return false;
+                    }
+
+                    bucket = (bucket + 2) & (table.length - 1);
+                }
+            } finally {
+                unlockWrite(stamp);
+            }
+        }
+
+        private boolean removeIf(Predicate<LongPair> filter) {
```

Comment:
Do not use `Predicate<LongPair>` because when we iterate we need to create all the `new LongPair()` instances.

Instead, you can define a: 

```java
interface LongPairPredicate {
    boolean test(long v1, long v2);
}
```